app/services/record_declarations/base.rb

```diff
@@ -106,9 +95,57 @@ def create_record!
       end
     end
 
+    def valid_courses_for_user
+      valid_courses = []
+      valid_courses << "ecf-mentor" if user.mentor?
+      valid_courses << "ecf-induction" if user.early_career_teacher?
+      valid_courses += NPQCourse.all.map(&:identifier) if user.npq?
+      valid_courses
+    end
+
+    def profile_exists
+      return if errors.any?
+
+      unless user_profile
+        errors.add(:user_profile, I18n.t(:invalid_participant))
+      end
+    end
+
+    def date_has_the_right_format
+      return if declaration_date.blank?
+
+      errors.add(:declaration_date, I18n.t(:invalid_declaration_date)) unless declaration_date.match(RCF3339_DATE_REGEX)
+      errors.add(:declaration_date, I18n.t(:future_declaration_date)) if parsed_date > Time.zone.now
+    rescue StandardError
+      errors.add(:declaration_date, I18n.t(:invalid_declaration_date))
+    end
+
     def validate_provider!
       # TODO: Remove the nil? check and fix the test setup so that they build the school cohort, partnership and give us back the actual lead_provider.
       raise ActionController::ParameterMissing, I18n.t(:invalid_participant) unless actual_lead_provider.nil? || lead_provider_from_token == actual_lead_provider
     end
+
+    def validate_schedule!
+      schedule = user_profile.schedule
```

Comment:
this is ripe for extracting to a method along with `existing_declarations` and `next_milestone` unless it is problematic?